python/pyarrow/tests/test_dataset.py

```diff
@@ -0,0 +1,364 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import pytest
+
+import pyarrow as pa
+import pyarrow.fs as fs
+
+try:
+    import pyarrow.dataset as ds
+except ImportError:
+    ds = None
+
+# Marks all of the tests in this module
+# Ignore these with pytest ... -m 'not dataset'
+pytestmark = pytest.mark.dataset
+
+
+@pytest.fixture
+@pytest.mark.parquet
+def mockfs(table):
+    import pyarrow.parquet as pq
+
+    mockfs = fs._MockFileSystem()
+
+    directories = [
+        'subdir/1/xxx',
+        'subdir/2/yyy',
+    ]
+    for i, directory in enumerate(directories):
+        path = '{}/file{}.parquet'.format(directory, i)
+        mockfs.create_dir(directory)
+        with mockfs.open_output_stream(path) as out:
+            pq.write_table(table, out)
+
+    return mockfs
+
+
+@pytest.fixture
+def schema():
+    return pa.schema([
+        pa.field('i64', pa.int64()),
+        pa.field('f64', pa.float64())
+    ])
+
+
+@pytest.fixture
+def record_batch(schema):
+    data = [
+        list(range(5)),
+        list(map(float, range(5)))
+    ]
+    return pa.record_batch(data, schema=schema)
+
+
+@pytest.fixture
+def table(record_batch):
+    return pa.Table.from_batches([record_batch])
+
+
+@pytest.fixture
+def simple_data_source(record_batch):
+    return ds.SimpleDataSource([record_batch])
+
+
+@pytest.fixture
+def tree_data_source(simple_data_source):
+    return ds.TreeDataSource([simple_data_source] * 2)
+
+
+@pytest.fixture
+def dataset(simple_data_source, tree_data_source, schema):
+    return ds.Dataset([simple_data_source, tree_data_source], schema)
+
+
+def test_scan_context():
+    context = ds.ScanContext()
+    assert isinstance(context.memory_pool, pa.MemoryPool)
+    context = ds.ScanContext(pa.default_memory_pool())
+    assert isinstance(context.memory_pool, pa.MemoryPool)
```

Comment:
Do you want to check that the one you pass in is passed on correctly to the ScanContext? I think you can use `logging_memory_pool` to check